api/layer/object_cache_test.go

```diff
@@ -0,0 +1,170 @@
+)
+
+func randID(t *testing.T) *object.ID {
+	id := object.NewID()
+	id.SetSHA256(randSHA256Checksum(t))
+
+	return id
+}
+
+func randSHA256Checksum(t *testing.T) (cs [sha256.Size]byte) {
+	_, err := rand.Read(cs[:])
+	require.NoError(t, err)
+
+	return
+}
+
+func TestTrimAfterObjectName(t *testing.T) {
+	var (
+		objects []*ObjectInfo
+		names   = []string{"b", "c", "d"}
+	)
+	for _, name := range names {
+		objects = append(objects, &ObjectInfo{Name: name})
+	}
+
+	t.Run("startafter before all objects", func(t *testing.T) {
+		actual := trimAfterObjectName("a", objects)
+		require.Equal(t, objects, actual)
+	})
+
+	t.Run("startafter first object", func(t *testing.T) {
+		actual := trimAfterObjectName(names[0], objects)
+		require.Equal(t, objects[1:], actual)
+	})
+
+	t.Run("startafter second-to-last object", func(t *testing.T) {
+		actual := trimAfterObjectName(names[len(names)-2], objects)
+		require.Equal(t, objects[len(objects)-1:], actual)
+	})
+
+	t.Run("startafter last object", func(t *testing.T) {
+		actual := trimAfterObjectName(names[len(names)-1], objects)
+		require.Empty(t, actual)
+	})
+
+	t.Run("startafter after all objects", func(t *testing.T) {
+		actual := trimAfterObjectName("z", objects)
+		require.Nil(t, actual)
+	})
+
+	t.Run("empty objects", func(t *testing.T) {
+		actual := trimAfterObjectName(names[0], []*ObjectInfo{})
+		require.Nil(t, actual)
+	})
+
+	t.Run("nil objects", func(t *testing.T) {
+		actual := trimAfterObjectName(names[0], nil)
+		require.Nil(t, actual)
+	})
+
+	t.Run("empty startafter", func(t *testing.T) {
+		actual := trimAfterObjectName("", objects)
+		require.Equal(t, actual, objects)
+	})
+}
+
+func TestTrimAfterObjectID(t *testing.T) {
+	var (
+		objects     []*ObjectInfo
+		ids         []*object.ID
+		numberOfIDS = 3
+	)
+
+	for i := 0; i < numberOfIDS; i++ {
+		id := randID(t)
+		objects = append(objects, &ObjectInfo{id: id})
+		ids = append(ids, id)
+	}
+
+	t.Run("existing id", func(t *testing.T) {
+		actual := trimAfterObjectID(ids[0].String(), objects)
+		require.Equal(t, objects[1:], actual)
+	})
+
+	t.Run("second to last id", func(t *testing.T) {
+		actual := trimAfterObjectID(ids[len(ids)-2].String(), objects)
+		require.Equal(t, objects[len(objects)-1:], actual)
+	})
+
+	t.Run("non-existing id", func(t *testing.T) {
+		actual := trimAfterObjectID("z", objects)
+		require.Nil(t, actual)
+	})
+
+	t.Run("last id", func(t *testing.T) {
+		actual := trimAfterObjectID(ids[len(ids)-1].String(), objects)
+		require.Nil(t, actual)
+	})
+
+	t.Run("empty id", func(t *testing.T) {
+		actual := trimAfterObjectID("", objects)
+		require.Nil(t, actual)
+	})
+}
+
+func TestObjectsListV2Cache(t *testing.T) {
+	var (
+		cacheSize = 10
+		objects   []*ObjectInfo
+		key       = "key"
+	)
+
+	for i := 0; i < cacheSize; i++ {
+		id := randID(t)
+		objects = append(objects, &ObjectInfo{id: id, Name: id.String()})
+	}
+
+	t.Run("lifetime", func(t *testing.T) {
+		cache := newListObjectsCache()
```

Comment:
Maybe add an optional parameter to define the cache lifetime? 60 seconds of waiting for a test is too long for me.

Reply:
We talked about this with @alexvanin, but we decided to leave cache lifetime hardcoded because there is no such need to add this as a parameter. 

But you're right this test takes too long. Maybe I'd be better to separate this test somehow?

Reply:
We can extract this functionality to separate function and test it. For example:
```
func (l *listObjectsCache) scheduleClearing(key string, d time.Duration) {
	time.AfterFunc(d, func() {
		l.mtx.Lock()
		delete(l.caches, key)
		l.mtx.Unlock()
	})
}
```

Reply:
> but we decided to leave cache lifetime hardcoded because there is no such need to add this as a parameter.

60 seconds can be hardcoded in terms of "not configurable from the S3 Gateway configuration". However the program component itself may be configured. Let's do that and make these tests pass faster.